src/engine/strat_engine/backstore/device.rs

```diff
@@ -43,3 +46,159 @@ pub fn resolve_devices<'a>(paths: &'a [&Path]) -> StratisResult<HashMap<Device,
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum DevOwnership {
+    Ours(PoolUuid, DevUuid),
+    Unowned,
+    Theirs(String), // String is something useful to give back to end user about what's on device
+}
+
+/// Returns true if a device has no signature, yes this is a bit convoluted.  Logic gleaned from
+/// blivet library.
+fn empty(device: &HashMap<String, String>) -> bool {
+    if (device.contains_key("ID_PART_TABLE_TYPE") && !device.contains_key("ID_PART_ENTRY_DISK"))
+        || device.contains_key("ID_FS_USAGE")
+    {
+        false
+    } else {
+        true
+    }
+}
+
+/// Generate some kind of human readable text about what's on a device.
+fn signature(device: &HashMap<String, String>) -> String {
+    if empty(device) {
+        String::from("empty")
+    } else {
+        device
+            .iter()
+            .filter(|&(k, _)| k.contains("ID_FS_") || k.contains("ID_PART_TABLE_"))
+            .map(|(k, v)| format!("{}={}", k, v))
+            .collect::<Vec<String>>()
+            .join(" ")
+    }
+}
+
+/// Determine what a block device is used for.
+pub fn identify(devnode: &Path) -> StratisResult<DevOwnership> {
+    let rc = if let Some(device) = get_udev_block_device(devnode)? {
+        if empty(&device) {
+            // The device is either really empty or we are running on a distribution that hasn't
+            // picked up the latest libblkid, lets read down to the device and find out for sure.
+            // TODO: At some point in the future we can remove this and just return Unowned.
+            if let Some((pool_uuid, device_uuid)) = StaticHeader::device_identifiers(
+                &mut OpenOptions::new().read(true).open(&devnode)?,
+            )? {
+                Ok(DevOwnership::Ours(pool_uuid, device_uuid))
+            } else {
+                Ok(DevOwnership::Unowned)
+            }
+        } else if device.contains_key("ID_FS_TYPE")
+            && device.get("ID_FS_TYPE").unwrap() == "stratis"
+        {
+            // Device is ours, but we don't get everything we need from udev db, lets go to disk.
+            if let Some((pool_uuid, device_uuid)) = StaticHeader::device_identifiers(
+                &mut OpenOptions::new().read(true).open(&devnode)?,
+            )? {
+                Ok(DevOwnership::Ours(pool_uuid, device_uuid))
+            } else {
+                // In this case the udev db says it's ours, but our check says otherwise.  We should
+                // trust ourselves.  Should we raise an error here?
+                Ok(DevOwnership::Theirs(String::from(
+                    "Udev db says stratis, disk meta says no",
+                )))
+            }
+        } else {
+            Ok(DevOwnership::Theirs(signature(&device)))
+        }
+    } else {
+        Err(StratisError::Engine(
+            ErrorEnum::NotFound,
+            String::from(format!(
+                "We expected to find the block device {:?} \
+                 in the udev db",
+                devnode
```

Comment:
With our current formatting, I think this string can all fit on one line.

Reply:
Done